### DISCUSSION TITLE
ShouldBeUnique causing jobs to fail tests

### DISCUSSION THREAD

**Discussion by ngunyimacharia** [laravel/framework#40233] [Q&A]
When running tests in Laravel, after mocking the queue using `Queue::fake();`, Laravel will still try to acquire the unique lock for the job. This causes failure of jobs as the lock is only released when the job is complete which never happens.

When the job with the same `uniqueId` is dispatched again the lock is not acquired thus the test isn't `pushed` to the Queue or `dispatched` to the `Bus`.

Is this the ideal behavior and is there a suggested workaround?

**Comment by alecpl** [comment]
I made one of my jobs `ShouldbeUnique` and tests started failing. So, I confirm this is a problem. `Queue::fake()` call should reset the queue manager state regarding job uniqueness. Or some other way to release the lock.

**Comment by alecpl** [comment]
Here's a way to remove lock for jobs in fake queue:
```php
        foreach (Queue::pushedJobs() as $list) {
            foreach ($list as $item) {
                if ($item['job'] instanceof \Illuminate\Contracts\Queue\ShouldBeUnique) {
                    $cache = $this->app->make(\Illuminate\Contracts\Cache\Repository::class);
                    (new \Illuminate\Bus\UniqueLock($cache))->release($item['job']);
                }
            }
        }
```
this however is not very useful. More useful would be to remove all existing locks. Note that `cache()->flush()`  does not work for me. I had to use `cache()->getStore()->lockConnection()->flushall();`. This however does not look like the best solution.

Here's another one (but only for redis)
```php
        $db = \cache()->getStore()->lockConnection();
        $prefix = $db->getOptions()->prefix?->getPrefix();

        foreach ($db->keys('*') as $key) {
            if (strpos($key, 'laravel_unique_job') !== false) {
                $db->del($prefix ? substr($key, strlen($prefix)) : $key);
            }
        }
```

**Comment by rubentebogt** [comment]
Did anyone find a good solutions for this? I'm having the same issue here @alecpl @ngunyimacharia ?

**Comment by frankiejarrett** [comment]
@rubentebogt I've been using this little workaround in my jobs that implement `ShouldBeUnique` for a few years now. Seems to work great and haven't really had any issues. Good luck!

```php
public function uniqueId(): string
{
    return app()->environment('testing') ? microtime() : $this->model->uuid;
}
```

**Comment by zikeji** [comment]
In my case I overrode the cache driver to be array for the specific test, since the actual job functionality is tested separately, this is only asserting it gets queued.
